### PR TITLE
Add the criteria for including new APIs

### PR DESCRIPTION
**Pull request by brendandburns**
@bgrant0607 @smarterclayton @davidopp @thockin @jbeda 

As discussed in community meeting today.

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -48,6 +48,37 @@ Kubernetes also stores its serialized state (currently in [etcd](https://coreos.
...
+   * The API object is expected to be generally useful to greater than 50% of the Kubernetes users.  This is to ensure that we don't build up a collection of niche APIs
+that users rarely need.
+   * There is general consensus in the Kubernetes community that the API object is in the "Kubernetes layer".  See ["What is Kubernetes?"](whatisk8s.md) for a detailed
+explanation of what we believe the "Kubernetes layer" to be.
+
+Of course for every set of rules, we need to ensure that we are not hamstrung or limited by slavish devotion to those rules, and thus we also include two exceptions
+for including APIs in Kubernetes that violate these criteria.
+
+These exceptions are:
+   * There is no other way to implement the functionality in KuberneteOf course for every set of rules, we need to ensure that we are not hamstrung or limited by slavish devotion to those rules, and thus we also include two exceptions
```
Comment (smarterclayton) [inline docs/api.md]: Typo

**Comment by smarterclayton** [comment]
Nice concise definition

File: docs/api.md [hunk 2]
```diff
@@ -48,6 +48,37 @@ Kubernetes also stores its serialized state (currently in [etcd](https://coreos.
...
 Kubernetes itself is decomposed into multiple components, which interact through its API.
 
+## Adding APIs to Kubernetes
+Every API that is added to Kubernetes carries with it increased cost and complexity for all parts of the Kubernetes ecosystem.  New APIs imply new code to maintain,
+new tests that may flake, new documentation that users are required to understand, increased cognitive load for kubectl users and many other incremental costs.
+
+Of course, the addition of new APIs also enables new functionality that empowers users to simply do things that may have been previously complex, costly or both.
+
+Given this balance between increasing the complexity of the project, vs the reduction of complexity in user actions, we have set out to set up a set of criteria
+by which we as a development community decide when an API should be added to the set of core Kubernetes APIs.
```
Comment (davidopp) [inline docs/api.md]: I would soften this a little, e.g. "we have set out to set up a set of criteria that will help to guide how we as a development community..."

File: docs/api.md (hunk 1 above)
Comment (brendandburns) [inline docs/api.md]: done.

File: docs/api.md (hunk 2 above)
Comment (brendandburns) [inline docs/api.md]: done.

**Comment by brendandburns** [comment]
comments addressed, please take a look.

**Comment by jbeda** [comment]
LGTM -- separately I'd love to see a more clear designation of who is responsible for judging this stuff.  Ideally it'll be consensus most of the time unless it isn't.  This brings in the "k8s elders" thing that we've discussed in the past.

**Comment by bgrant0607** [comment]
LGTM, thanks.

**Comment by bgrant0607** [comment]
Please run hack/update-generated-docs.sh.

Feel free to reapply lgtm.

File: docs/api.md [hunk 3]
```diff
@@ -48,6 +48,33 @@ Kubernetes also stores its serialized state (currently in [etcd](https://coreos.
 
 Kubernetes itself is decomposed into multiple components, which interact through its API.
 
+## Adding APIs to Kubernetes
+Every API that is added to Kubernetes carries with it increased cost and complexity for all parts of the Kubernetes ecosystem.  New APIs imply new code to maintain,
+new tests that may flake, new documentation that users are required to understand, increased cognitive load for kubectl users and many other incremental costs.
+
+Of course, the addition of new APIs also enables new functionality that empowers users to simply do things that may have been previously complex, costly or both.
+
+Given this balance between increasing the complexity of the project, vs the reduction of complexity in user actions, we have set out to set up a set of criteria
```
Comment (bgrant0607) [inline docs/api.md]: Nit: Remove the comma prior to "vs".

**Comment by wojtek-t** [comment]
@brendandburns 

FAIL: changes needed but not made due to --verify
/go/src/k8s.io/kubernetes/docs/ is out of date. Please run hack/update-generated-docs.sh
FAILED   ./hack/../hack/verify-generated-docs.sh	62s

Please fix

File: docs/api.md (hunk 3 above)
Comment (brendandburns) [inline docs/api.md]: done.

**Comment by brendandburns** [comment]
generated docs. addressed comment. re-applying lgtm.

**Comment by brendandburns** [comment]
@k8s-bot unit test this issue: #26721

**Comment by brendandburns** [comment]
@bgrant0607 this won't merge b/c it's not marked 1.3.  Should I mark it 1.3 or 1.4?

**Comment by bgrant0607** [comment]
Marking 1.3 so it can be merged.

**Pull request merged**